### PR TITLE
update openssl to 3 series

### PR DESCRIPTION
**Pull request by dkwo**
[ci skip]

Note to self:
- use `xrevbump "bump for openssl-3.0" $(xbps-query -Rs libssl.so -p shlib-requires | cut -d ':' -f 1 | xargs -n1 xbps-uhelper getpkgname)` and same for `libcrypto`, or use `sort -u file1 file2`.

### CODE REVIEW

**Comment by paper42** [comment]
I am pretty sure not all packages work with openssl 3, converting to draft

**Comment by dkwo** [comment]
@Duncaen Do you know how to fix the `xbps` config failure?

**Comment by leahneukirchen** [comment]
Please delay until 3.0.5.

**Comment by dkwo** [comment]
This is an incomplete list of build failures (could be outdated):
- gsoap, vanitygen: missing distfiles
- tcpflow: wrong sha-sums
- cbang, poco, pidgin-sipe, nodejs-lts, kore, xrootd, netpgp, opensc, vpn-ws, eid-mw, boinc: deprecated declarations
- xca: deprecated-copy
- mairix, rust-sccache, fractal, telegram-desktop, rabbitmq-c, git-brws, rooster, cargo-aoc, spotify-tui: weird errors
- php: discarded-qualifiers
- dmg2img: openssl patch
- lua54-luasec: error
- yubiko-piv-tool: unknown type name errors
- impi-util: error

**Comment by dkwo** [comment]
After rebuilding `xbps`, it complains that the RSA signature is not valid, and refuses to install packages.
Any hint how to deal with this?

**Comment by dkwo** [comment]
Plan to revisit this at some point.

**Comment by dkwo** [comment]
Actually, xbps fails one test:
```
===> Failed tests
libxbps/shell/obsoletefiles_test:replace_package_same_files  ->  failed: 0 != 2 (0 != 2)  [0.260s]
```

**Review by paper42 — COMMENTED** [review]
I would also keep libssl1.1 and libcrypto1.1 packaged separately for backwards compatibility

File: srcpkgs/openssl/template [hunk 1]
```diff
@@ -58,36 +58,26 @@ pre_check() {
...
 	}
 }
 
-openssl-c_rehash_package() {
-	short_desc+=" - c_rehash utility"
-	depends="openssl perl"
-	pkg_install() {
-		vmove usr/bin/c_rehash
-	}
-}
```
[the remark this answers is not in the document]
Comment (dkwo) [reply srcpkgs/openssl/template:87]: Sorry, I meant to remove the script. I'll just keep it packaged separately then.

File: srcpkgs/openssl/template [hunk 2]
```diff
@@ -58,36 +58,26 @@ pre_check() {
...
 
 post_install() {
-	if [ ! "$CHROOT_READY" ]; then
-		rm -f "${DESTDIR}/usr/bin/c_rehash"
-	fi
+	rm -f "${DESTDIR}/usr/bin/c_rehash"
 }
 
-libcrypto1.1_package() {
+libcrypto_package() {
```
[the remark this answers is not in the document]
Comment (dkwo) [reply srcpkgs/openssl/template:64]: restored it.

**Comment by dkwo** [comment]
> I would also keep libssl1.1 and libcrypto1.1 packaged separately for backwards compatibility

Should I then keep an `openssl1.1-compat` package or similar?

File: srcpkgs/openssl/template [hunk 3]
```diff
@@ -26,11 +26,11 @@ build_options=asm
 if [ "$CHROOT_READY" ]; then
 	hostmakedepends="perl"
 	build_options_default="asm"
-else
-	# openssl-c_rehash depends on perl, ignore on bootstrap
-	subpackages="libcrypto1.1 libssl1.1 openssl-devel"
 fi
 
+# due to our libatomic issues
+makedepends+=" libatomic-devel"
```
[the remark this answers is not in the document]
Comment (dkwo) [reply srcpkgs/openssl/template:32]: done.

**Comment by paper42** [comment]
> > I would also keep libssl1.1 and libcrypto1.1 packaged separately for backwards compatibility
> 
> Should I then keep an `openssl1.1-compat` package or similar?

I might be wrong, but I think we only need libssl1.1 and libcrypto1.1, not what is in the openssl package, so it should be a package called libssl1.1 with a subpackage libcrypto1.1 I think.

**Comment by paper42** [comment]
most of these changes look like they could be merged separately, so please open separate PRs for them

**Comment by dkwo** [comment]
Besides xbps, the remaining failures are `git-brws` @jcgruenhage , `git-series` @leahneukirchen , `taisei` @SolitudeSF , `signal-backup-decode` @sburris0 , `gcsf` @benalb : these are packages that have not seen updates upstream, for several years in some cases. Can anyone help?

**Comment by jcgruenhage** [comment]
@dkwo I haven't finished trying that yet, but I'd guess that doing this at some point prior to the build (post_patch probably makes the most sense here) would fix at least the rust ones:

```
	# required for openssl3 compat
	cargo update --package openssl-sys:0.9.75 --precise 0.9.80
	cargo update --package openssl:0.10.41 --precise 0.10.45
```

The first version is what the lockfile currently contains, the second is the version we're bumping to, which in the example above is the current latest.

**Comment by leahneukirchen** [comment]
git-series can be dropped if necessary.

**Comment by benalb** [comment]
> gcsf
I still use gcsf now and then, but upstream has no interest in the project, so it can be dropped, I will keep a copy my repo.

**Comment by jcgruenhage** [comment]
Okay, so there's a bit more to git-brws in specific, but with this diff it now builds based on your branch:

```
diff --git a/srcpkgs/git-brws/template b/srcpkgs/git-brws/template
index ea5e550b01..9ebcec181a 100644
--- a/srcpkgs/git-brws/template
+++ b/srcpkgs/git-brws/template
@@ -1,7 +1,7 @@
 # Template file for 'git-brws'
 pkgname=git-brws
 version=0.11.12
-revision=2
+revision=3
 build_style=cargo
 hostmakedepends="pkg-config"
 makedepends="openssl-devel"
@@ -12,9 +12,16 @@ homepage="https://rhysd.github.io/git-brws/"
 distfiles="https://github.com/rhysd/git-brws/archive/v${version}.tar.gz"
 checksum=3a4bbf93f0b16562260ca66c2b60c655d5bfc690d0229d11757be76d95cb81c5
 
-pre_build() {
+post_patch() {
        # fixes an indexmap error when cross compiling
        cargo update --package autocfg:1.0.1 --precise 1.1.0
+
+       # required for openssl3 compat
+       cargo update --package openssl-sys:0.9.58 --precise 0.9.80
+       cargo update --package openssl:0.10.30 --precise 0.10.45
+
+       # required for compatibility with Rust >= 1.64.0
+       cargo update --package socket2:0.3.15 --precise 0.3.19
 }
 
 post_install() {
 ```
 
 Want me to commit that to a branch and push it somewhere so that you can cherry-pick, or do you want to do that yourself?

**Comment by jcgruenhage** [comment]
I've looked into the packages mentioned above a bit and can report:

- signal-backup-decode, git-series and git-brws are easy fixes, I've pushed my locally built patches to https://github.com/jcgruenhage/void-packages/commits/openssl3.
- gcsf doesn't build even with openssl/openssl-sys bumped to a newer version, and I haven't looked into this further
- taisei I didn't look into further after I saw it's not a rust package.

**Comment by dkwo** [comment]
> So that means the only remaining known problem is with xbps, right?

Correct.

**Comment by dkwo** [comment]
Actually, `taisei` may work with a small fix, so I may go back to it later.

**Comment by dkwo** [comment]
@jcgruenhage For `git-brws`, on `x86_64-musl` I get `test result: FAILED. 78 passed; 25 failed`, not sure how it was before, but if you don't complain I'll leave it like that for now :)

**Comment by jcgruenhage** [comment]
I wasn't running tests on the openssl build, so I can't say, but I'm pretty sure that the tests used to work at some point..

**Comment by dkwo** [comment]
I get exactly the same on the master branch: `test result: FAILED. 78 passed; 25 failed`.

**Comment by paper42** [comment]
Please add a package providing libssl1.1, libcrypto1.1 and openssl1.1-devel

**Comment by dkwo** [comment]
@paper42 Done. I might as well use it for `gcfs`, instead of removing it.

**Comment by dkwo** [comment]
@leahneukirchen Should I let `racket` still depend on `libssl1.1` instead of 3.0 ?

**Comment by dkwo** [comment]
I'll mark this for review, with the caveat that `xbps` is still broken, there are a few PR's still pending (linked here) and I will need to revbump everything later.

**Comment by leahneukirchen** [comment]
On Wed, Feb 22, 2023, 22:04 dkwo ***@***.***> wrote:

> @leahneukirchen <https://github.com/leahneukirchen> Should I let racket
> still depend on libssl1.1 instead of 3.0 ?
>
No, 3.0 should work:
https://github.com/racket/racket/commit/b15f33b0bc3aca54c3fdc2bcd59f740cbb6fcb6f

**Comment by jcgruenhage** [comment]
@dkwo if xbps is currently broken with OpenSSL 3.0, shouldn't we just build it against openssl1.1-devel for now?

**Comment by jcgruenhage** [comment]
@dkwo if xbps is currently broken with OpenSSL 3.0, shouldn't we just build it against openssl1.1-devel for now?

**Comment by paper42** [comment]
> @dkwo if xbps is currently broken with OpenSSL 3.0, shouldn't we just build it against openssl1.1-devel for now?

We can, but that won't make this merge-able. Having both libssl1.1 and 3.0 as a dependency of base-system would be very ugly.

**Comment by dkwo** [comment]
It seems we do not need to skip tests in openssl.

File: srcpkgs/openssl1.1-compat/template [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# Template file for 'openssl1.1-compat'
+pkgname=openssl1.1-compat
```
Comment (paper42) [inline srcpkgs/openssl1.1-compat/template:2]: this definitely $conflicts with openssl

File: srcpkgs/openssl1.1-compat/template [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+libssl1.1_package() {
+	short_desc+=" - SSL/TLS library"
+	pkg_install() {
+		vmove "usr/lib/libssl.so.*"
+	}
+}
+
+openssl1.1-devel_package() {
```
Comment (paper42) [inline srcpkgs/openssl1.1-compat/template:67]: this $conflicts with openssl-devel

Comment (dkwo) [reply srcpkgs/openssl1.1-compat/template:68]: done.

File: srcpkgs/openssl1.1-compat/template (hunk 4 above)
Comment (dkwo) [reply srcpkgs/openssl1.1-compat/template:2]: done, thanks

**Comment by dkwo** [comment]
> I might as well use it for `gcfs`, instead of removing it.

Actually, it fails to build due to `conflicts`, so I'll remove `gcfs`.

File: srcpkgs/taisei/template [hunk 6]
```diff
@@ -13,6 +13,8 @@ license="MIT"
 homepage="https://taisei-project.org"
 distfiles="https://github.com/taisei-project/taisei/releases/download/v${version}/taisei-v${version}.tar.xz"
 checksum=dbc05f1b5c31981d8711130ac283355b7bfad403895f4096a6bb7e9a3d73a3bc
+# build failure with openssl 3
+CFLAGS+=" -Wno-error"
```
Comment (paper42) [inline srcpkgs/taisei/template:17]: import this patch instead https://github.com/taisei-project/taisei/pull/344

Comment (dkwo) [reply srcpkgs/taisei/template:17]: done. that patch alone is not enough, as the released tarball does not include previous changes, even if they were made a year before its release. i added both

**Comment by dkwo** [comment]
should i update this to 3.1 or stick with 3.0 lts?

**Comment by dkwo** [comment]
i updated the status in first comment: there are a couple of packages that do not cross compile, and one waiting for upstream release that works with musl without patches. xbps is the major issue. the rest are simple updates.

**Comment by dkwo** [comment]
Once `gcsf` is updated #45250 nothing is using openssl-compat.
I'm leaving it just in case we want to build `xbps` with it (the ugly way).

**Comment by classabbyamp** [comment]
3.1.2 is available

**Comment by classabbyamp** [comment]
php8.0 ossl3 patchs can be found at https://git.alpinelinux.org/aports/tree/testing/php8?h=master

**Comment by classabbyamp** [comment]
I'll have some patches for xbps soon

**Comment by dkwo** [comment]
@classabbyamp Indeed, should we update to 3.1 series, or stay with 3.0, which is LTS?
as for php8.0, I tried to do that, but then consesus seemed to be remove it, see https://github.com/void-linux/void-packages/pull/42231#issuecomment-1458283188
Great that you can patch xbps, that's the most important thing.

**Comment by classabbyamp** [comment]
ok fair enough wrt php8

**Comment by dkwo** [comment]
@classabbyamp With your recent changes to `xbps`, it only fails this test:
```
===> Failed tests
libxbps/shell/obsoletefiles_test:replace_package_same_files  ->  failed: 0 != 2 (0 != 2)  [0.196s]
```
but this also fails on the master branch, so it's likely unrelated to this update.
Apart from that, it seems to work fine.

**Comment by classabbyamp** [comment]
hm didn't mean to push that to this PR, not sure how that happened

**Comment by dkwo** [comment]
Sure. If you find any failure that is openssl-related, feel free to push it here.

**Pull request closed without merging**